src/oob/idcert.rs

```diff
@@ -0,0 +1,526 @@
+        update_once(basic_ca, || {
+            match cons.take_sequence(|cons| cons.take_opt_bool())? {
+                Some(res) => Ok(res),
+                None => Ok(false)
+            }
+        })
+    }
+
+    /// Parses the Subject Key Identifier Extension.
+    ///
+    /// The extension must be present and contain the 160 bit SHA-1 hash of
+    /// the value of the DER-encoded bit string of the subject public key.
+    ///
+    /// ```text
+    /// SubjectKeyIdentifier ::= KeyIdentifier
+    /// KeyIdentifier        ::= OCTET STRING
+    /// ```
+    fn take_subject_key_identifier<S: Source>(
+        cons: &mut Constructed<S>,
+        subject_key_id: &mut Option<OctetString>
+    ) -> Result<(), S::Err> {
+        update_once(subject_key_id, || {
+            let id = OctetString::take_from(cons)?;
+            if id.len() != 20 {
+                xerr!(Err(Error::Malformed.into()))
+            } else {
+                Ok(id)
+            }
+        })
+    }
+
+    /// Parses the Authority Key Identifier Extension.
+    ///
+    /// Must be present except in self-signed CA certificates where it is
+    /// optional.
+    ///
+    /// ```text
+    /// AuthorityKeyIdentifier ::= SEQUENCE {
+    ///   keyIdentifier             [0] KeyIdentifier           OPTIONAL,
+    ///   authorityCertIssuer       [1] GeneralNames            OPTIONAL,
+    ///   authorityCertSerialNumber [2] CertificateSerialNumber OPTIONAL  }
+    ///
+    /// KeyIdentifier ::= OCTET STRING
+    /// ```
+    ///
+    /// Only keyIdentifier must be present.
+    fn take_authority_key_identifier<S: Source>(
+        cons: &mut Constructed<S>,
+        authority_key_id: &mut Option<OctetString>
+    ) -> Result<(), S::Err> {
+        update_once(authority_key_id, || {
+            let res = cons.take_sequence(|cons| {
+                cons.take_value_if(Tag::CTX_0, OctetString::take_content_from)
+            })?;
+            if res.len() != 20 {
+                return Err(Error::Malformed.into())
+            } else {
+                Ok(res)
+            }
+        })
+    }
+}
+
+
+//------------ OIDs ----------------------------------------------------------
+
+mod oid {
+    use ::ber::Oid;
+
+    pub const RSA_ENCRYPTION: Oid<&[u8]>
+    = Oid(&[42, 134, 72, 134, 247, 13, 1, 1, 1]);
+
+    pub const CE_SUBJECT_KEY_IDENTIFIER: Oid<&[u8]> = Oid(&[85, 29, 14]);
+    pub const CE_BASIC_CONSTRAINTS: Oid<&[u8]> = Oid(&[85, 29, 19]);
+    pub const CE_AUTHORITY_KEY_IDENTIFIER: Oid<&[u8]> = Oid(&[85, 29, 35]);
+}
+
+
+//------------ Tests ---------------------------------------------------------
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use std::fs;
+    use bytes::Bytes;
+    use time;
+    use chrono::{TimeZone, Utc};
+
+    #[test]
+    fn test_parse_id_publisher_ta_cert() {
+        let d = Utc.ymd(2012, 1, 1).and_hms(0, 0, 0);
+        time::with_now(d, || {
+        let data = fs::read("test/oob/id-publisher-ta.cer").unwrap();
```

Comment:
Use the `input!` macro instead of doing an actual file system read?

Reply:
done (also in exchange.rs)